### PR TITLE
Use unified group creation wizard

### PR DESCRIPTION
**Pull request by DonOmalVindula**
### Purpose
$subject

### Related Issues
- https://github.com/wso2/product-is/issues/18137
- https://github.com/wso2/product-is/issues/18140

### Related PRs
- Related PR `#1` or (None)

### Checklist
- [ ] e2e cypress tests locally verified.
- [ ] Manual test round performed and verified.
- [ ] UX/UI review done on the final implementation.
- [ ] Documentation provided. (Add links if there are any)
- [ ] Unit tests provided. (Add links if there are any)
- [ ] Integration tests provided. (Add links if there are any)

### Security checks
- [ ] Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines?
- [ ] Ran FindSecurityBugs plugin and verified report?
- [ ] Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets?

### CODE REVIEW

File: apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx [hunk 1]
```diff
@@ -0,0 +1,493 @@
...
+import { getGroupsWizardStepIcons } from "../../configs";
+import {
+    CreateGroupInterface,
+    CreateGroupMemberInterface
+} from "../../models";
+
+/**
+ * Interface which captures create group props.
+ */
+interface CreateGroupProps extends TestableComponentInterface {
```
Comment (pavinduLakshan) [inline apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx:47]: Any reason not to use the IdentifiableComponentInterface?

File: apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx [hunk 2]
```diff
@@ -0,0 +1,493 @@
...
+
+/**
+ * Interface for wizard step.
+ */
+interface WizardStepInterface {
+    content: ReactElement;
+    icon: GenericIconProps | any;
+    name: string;
+    title: string;
+}
```
Comment (pavinduLakshan) [inline apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx:80]: Shall we move this to models folder?

File: apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx [hunk 3]
```diff
@@ -0,0 +1,493 @@
...
+ * @param props - props related to the create group wizard Member
+ */
+export const CreateGroupWizardUpdated: FunctionComponent<CreateGroupProps> =
+(props: CreateGroupProps): ReactElement => {
+
+    const {
+        closeWizard,
+        initStep,
+        requiredSteps,
+        [ "data-testid" ]: testId
```
Comment (pavinduLakshan) [inline apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx:94]: Let's use data-componentid instead.

File: apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx [hunk 4]
```diff
@@ -0,0 +1,493 @@
...
+        setCurrentWizardStep(currentStep - 1);
+        setPartiallyCompletedStep(undefined);
+    }, [ partiallyCompletedStep ]);
+
+    /**
+     * Method to handle create group action when create group wizard finish action is triggered.
+     *
+     * @param groupDetails - basic data required to create group.
+     */
+    const addGroup = (groupDetails: any): void => {
```
Comment (pavinduLakshan) [inline apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx:138]: Shall we use the proper type here? Please check and fix the other places as well.

File: apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx [hunk 5]
```diff
@@ -0,0 +1,493 @@
...
+        let groupName: string = "";
+
+        selectedUserStore !== "primary"
+            ? groupName = groupDetails?.basic.basicDetails
+                ? groupDetails?.basic?.basicDetails?.domain + "/" + groupDetails?.basic?.basicDetails?.groupName
+                : groupDetails?.domain + "/" + groupDetails?.groupName
+            : groupName = groupDetails?.basic?.basicDetails
+                ? groupDetails?.basic?.basicDetails?.groupName
+                : groupDetails?.groupName;
```
Comment (pavinduLakshan) [inline apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx:147]: It'd be better if we can simplify nested ternary expressions into if else statements for better readability.

File: apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx [hunk 6]
```diff
@@ -0,0 +1,493 @@
...
+                            "groups": [ {
+                                "display": createdGroup.displayName,
+                                "value": createdGroup.id
+                            } ]
+                        }
+                    } ],
+                    "schemas": [ "urn:ietf:params:scim:api:messages:2.0:PatchOp" ]
+                };
+
+                if (rolesList && rolesList.length > 0) {
```
Comment (pavinduLakshan) [inline apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx:199]:
```suggestion
                if (rolesList?.length > 0) {
```

File: apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx [hunk 7]
```diff
@@ -0,0 +1,493 @@
...
+            }
+        }).finally(() => {
+            setIsSubmitting(false);
+        })
+        ;
+    };
+
+    /**
+     * Method to handle the create group wizard finish action.
+     *
```
Comment (pavinduLakshan) [inline apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx:286]:
```suggestion
```

File: apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx [hunk 8]
```diff
@@ -0,0 +1,493 @@
...
+     * Handles wizard step submit.
+     *
+     * @param values - Forms values to be stored in state.
+     * @param formType - Type of the form.
+     */
+    const handleWizardSubmit = (values: any, formType: WizardStepsFormTypes) => {
+        if (WizardStepsFormTypes.BASIC_DETAILS === formType) {
+            setWizardState({ ...wizardState, [ formType ]: values });
+            handleGroupWizardFinish({ ...wizardState, [ formType ]: values });
+
```
Comment (pavinduLakshan) [inline apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx:304]:
```suggestion
```

File: apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx [hunk 9]
```diff
@@ -0,0 +1,493 @@
...
+        setPartiallyCompletedStep(currentStep);
+    };
+
+    const handleFinishFlow = () => {
+        setSubmitGeneralSettings();
+    };
+
+    /**
+     * Filters the steps evaluating the requested steps.
+     *
```
Comment (pavinduLakshan) [inline apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx:330]:
```suggestion
```

File: apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx [hunk 10]
```diff
@@ -0,0 +1,493 @@
...
+     */
+    const filterSteps = (steps: WizardStepsFormTypes[]): WizardStepInterface[] => {
+
+        const getStepContent = (stepsToFilter: WizardStepsFormTypes[] | string[]) => {
+
+            const filteredSteps: any[] = [];
```
Comment (pavinduLakshan) [inline apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx:336]:
```suggestion
            const filteredSteps: WizardStepsFormTypes[] = [];
```

File: apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx [hunk 11]
```diff
@@ -0,0 +1,493 @@
...
+        if (!requiredSteps) {
+            return getStepContent(steps);
+        }
+
+        return getStepContent(intersection(steps, requiredSteps));
+    };
+
+    /**
+     * Get the wizard basic step.
+     *
```
Comment (pavinduLakshan) [inline apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx:356]: let's add `@returns` TSDoc tag here to state what is returned from this component.

File: apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx [hunk 12]
```diff
@@ -0,0 +1,493 @@
...
+        };
+    };
+
+    /**
+     * Resolves the step content.
+     *
+     * @returns step content.
+     */
+    const resolveStepContent = (): ReactElement => {
+
```
Comment (pavinduLakshan) [inline apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx:385]:
```suggestion
```

File: apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx [hunk 13]
```diff
@@ -0,0 +1,493 @@
...
+        switch (wizardSteps[currentStep]?.name) {
+            case WizardStepsFormTypes.BASIC_DETAILS:
+                return getBasicDetailsWizardStep()?.content;
+        }
+    };
+
+    return (
+        <Modal
+            open={ true }
```
Comment (pavinduLakshan) [inline apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx:398]:
```suggestion
            open
```

File: apps/console/src/features/groups/components/wizard/group-assign-users-updated.tsx [hunk 14]
```diff
@@ -0,0 +1,343 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { TestableComponentInterface } from "@wso2is/core/models";
```
Comment (pavinduLakshan) [inline apps/console/src/features/groups/components/wizard/group-assign-users-updated.tsx:19]: Let's use identifiablecomponentinterface here

File: apps/console/src/features/groups/components/wizard/group-assign-users-updated.tsx [hunk 15]
```diff
@@ -0,0 +1,343 @@
...
+    const [ listItemLimit, setListItemLimit ] = useState<number>(0);
+    const [ userListMetaContent, setUserListMetaContent ] = useState(undefined);
+    const [ listOffset ] = useState<number>(0);
+    const [ isUsersFetchRequestLoading, setIsUsersFetchRequestLoading ] = useState<boolean>(true);
+
+    const [ isSelectAllAssignedUsers ] = useState<boolean>(false);
+
+    const [ checkedAssignedListItems, setCheckedAssignedListItems ] = useState<UserBasicInterface[]>([]);
+
+
```
Comment (pavinduLakshan) [inline apps/console/src/features/groups/components/wizard/group-assign-users-updated.tsx:86]:
```suggestion
```

File: apps/console/src/features/groups/components/wizard/group-basics-updated.tsx [hunk 16]
```diff
@@ -0,0 +1,293 @@
...
+     *
+     * @param values - contains values from form elements
+     */
+    const handleDomainChange = (values: Map<string, FormValue>) => {
+        const domain: string = values.get("domain").toString();
+
+        setUserStore(domain);
+    };
+
+    /**
```
Comment (pavinduLakshan) [inline apps/console/src/features/groups/components/wizard/group-basics-updated.tsx:103]:
```suggestion
```

File: apps/console/src/features/groups/components/wizard/group-basics-updated.tsx [hunk 17]
```diff
@@ -0,0 +1,293 @@
...
+     */
+    const getUserStores = () => {
+        const storeOptions: DropdownItemProps[] = [
+            {
+                key: -1,
+                text: "Primary",
+                value: "primary"
+            }
+        ];
+        let storeOption: DropdownItemProps = {
```
Comment (pavinduLakshan) [inline apps/console/src/features/groups/components/wizard/group-basics-updated.tsx:151]:
```suggestion

        let storeOption: DropdownItemProps = {
```

File: apps/console/src/features/groups/components/wizard/group-basics-updated.tsx [hunk 18]
```diff
@@ -0,0 +1,293 @@
...
+
+                            } }
+                            value={ initialValues && initialValues.groupName }
+                            loading={ isRegExLoading }
+                        />
+                        <Hint>
+                            A name for the group.
+                            { " " }
+                            Can contain between 3 to 30 alphanumeric characters, dashes (<Code>-</Code>),{ " " }
+                            and underscores (<Code>_</Code>).
```
Comment (pavinduLakshan) [inline apps/console/src/features/groups/components/wizard/group-basics-updated.tsx:286]: Let's i18n this content

File: apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx (hunk 4 above)
Comment (DonOmalVindula) [reply apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx]: Fixed with d08680252dc6b5e81e1864daef86e3756305e27d

File: apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx (hunk 5 above)
Comment (DonOmalVindula) [reply apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx]: Fixed with d08680252dc6b5e81e1864daef86e3756305e27d

File: apps/console/src/features/groups/components/wizard/group-assign-users-updated.tsx (hunk 14 above)
Comment (DonOmalVindula) [reply apps/console/src/features/groups/components/wizard/group-assign-users-updated.tsx]: Fixed with d08680252dc6b5e81e1864daef86e3756305e27d

File: apps/console/src/features/groups/components/wizard/group-assign-users-updated.tsx (hunk 15 above)
Comment (DonOmalVindula) [reply apps/console/src/features/groups/components/wizard/group-assign-users-updated.tsx]: Fixed with d08680252dc6b5e81e1864daef86e3756305e27d

File: apps/console/src/features/groups/components/wizard/group-basics-updated.tsx (hunk 16 above)
Comment (DonOmalVindula) [reply apps/console/src/features/groups/components/wizard/group-basics-updated.tsx:104]: Fixed with d08680252dc6b5e81e1864daef86e3756305e27d

File: apps/console/src/features/groups/components/wizard/group-basics-updated.tsx (hunk 17 above)
Comment (DonOmalVindula) [reply apps/console/src/features/groups/components/wizard/group-basics-updated.tsx:152]: Fixed with d08680252dc6b5e81e1864daef86e3756305e27d

File: apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx (hunk 6 above)
Comment (DonOmalVindula) [reply apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx]: Fixed with d08680252dc6b5e81e1864daef86e3756305e27d

File: apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx (hunk 7 above)
Comment (DonOmalVindula) [reply apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx]: Fixed with d08680252dc6b5e81e1864daef86e3756305e27d

File: apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx (hunk 8 above)
Comment (DonOmalVindula) [reply apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx]: Fixed with d08680252dc6b5e81e1864daef86e3756305e27d

File: apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx (hunk 9 above)
Comment (DonOmalVindula) [reply apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx]: Fixed with d08680252dc6b5e81e1864daef86e3756305e27d

File: apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx (hunk 10 above)
Comment (DonOmalVindula) [reply apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx]: Fixed with d08680252dc6b5e81e1864daef86e3756305e27d

File: apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx (hunk 11 above)
Comment (DonOmalVindula) [reply apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx:333]: Fixed with d08680252dc6b5e81e1864daef86e3756305e27d

File: apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx (hunk 12 above)
Comment (DonOmalVindula) [reply apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx]: Fixed with d08680252dc6b5e81e1864daef86e3756305e27d

File: apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx (hunk 13 above)
Comment (DonOmalVindula) [reply apps/console/src/features/groups/components/wizard/create-group-wizard-updated.tsx]: Fixed with d08680252dc6b5e81e1864daef86e3756305e27d

File: apps/console/src/features/groups/components/wizard/group-basics-updated.tsx (hunk 18 above)
Comment (DonOmalVindula) [reply apps/console/src/features/groups/components/wizard/group-basics-updated.tsx:287]: Let's fix the i18n in a follow up PR

**Review by pavinduLakshan — APPROVED** [review]

File: apps/console/src/features/groups/components/wizard/group-assign-users-updated.tsx [hunk 19]
```diff
@@ -0,0 +1,331 @@
...
+            [ "name", "name" ],
+            [ "userName", "userName" ],
+            [ "id", "" ],
+            [ "profileUrl", "profileUrl" ],
+            [ "meta.lastModified", "meta.lastModified" ],
+            [ "meta.created", "" ]
+        ]));
+    }, []);
+
+
```
Comment (pavinduLakshan) [inline apps/console/src/features/groups/components/wizard/group-assign-users-updated.tsx:100]:
```suggestion
```

File: apps/console/src/features/groups/components/wizard/group-assign-users-updated.tsx [hunk 20]
```diff
@@ -0,0 +1,331 @@
...
+        setIsUsersFetchRequestLoading(true);
+
+        getUsersList(limit, offset, filter, null, userStore)
+            .then((response: UserListInterface) => {
+                // Exclude JIT users.
+                const responseUsers: UserBasicInterface[] = response?.Resources?.filter(
+                    (user: UserBasicInterface) => !user[ SCIMConfigs.scim.enterpriseSchema ]?.userSourceId);
+
+                if (responseUsers) {
+
```
Comment (pavinduLakshan) [inline apps/console/src/features/groups/components/wizard/group-assign-users-updated.tsx:134]:
```suggestion
```

File: apps/console/src/features/groups/components/wizard/group-assign-users-updated.tsx [hunk 21]
```diff
@@ -0,0 +1,331 @@
...
+    const handleSearchFieldChange = (e: FormEvent<HTMLInputElement>, query: string, list: UserBasicInterface[],
+        stateAction: Dispatch<SetStateAction<any>>) => {
+
+        let isMatch: boolean = false;
+        const filteredRoleList: UserBasicInterface[] = [];
+
+        if (!isEmpty(query)) {
+            const regExp: RegExp = new RegExp(escapeRegExp(query), "i");
+
+            list && list.map((user: UserBasicInterface) => {
```
Comment (pavinduLakshan) [inline apps/console/src/features/groups/components/wizard/group-assign-users-updated.tsx:195]:
```suggestion
            list?.map((user: UserBasicInterface) => {
```

File: apps/console/src/features/groups/components/wizard/group-assign-users-updated.tsx [hunk 22]
```diff
@@ -0,0 +1,331 @@
...
+
+    return (
+        <>
+            <GroupBasicsUpdated
+                data-componentid="add-group-form"
+                triggerSubmit={ triggerSubmit }
+                initialValues={ initialValues?.basic }
+                userStore={ selectedUserStore }
+                setUserStore={ setSelectedUserStore }
+                onSubmit={ (values: any) => {
```
Comment (pavinduLakshan) [inline apps/console/src/features/groups/components/wizard/group-assign-users-updated.tsx:242]: can't we use Map<string,FormValue> here?

File: apps/console/src/features/groups/components/wizard/group-assign-users-updated.tsx [hunk 23]
```diff
@@ -0,0 +1,331 @@
...
+                        users: checkedAssignedListItems
+                    });
+                } }
+            />
+            { isUsersFetchRequestLoading
+                ? <Segment basic><ContentLoader className="p-3" active /></Segment>
+                : initialUserList?.length > 0 &&
+                (
+                    <>
+                        <Heading as="h5" className="mt-3">Add Users</Heading>
```
Comment (pavinduLakshan) [inline apps/console/src/features/groups/components/wizard/group-assign-users-updated.tsx:254]: Let's i18n the strings

File: apps/console/src/features/groups/components/wizard/group-basics-updated.tsx [hunk 24]
```diff
@@ -0,0 +1,285 @@
...
+    triggerSubmit: boolean;
+    initialValues: any;
+    onSubmit: (values: any) => void;
+    userStore: string;
+    setUserStore: (userStore: string) => void;
+}
+
+/**
+ * Component to capture basic details of a new role.
+ *
```
Comment (pavinduLakshan) [inline apps/console/src/features/groups/components/wizard/group-basics-updated.tsx:50]:
```suggestion
```

File: apps/console/src/features/groups/components/wizard/group-assign-users-updated.tsx (hunk 22 above)
Comment (DonOmalVindula) [reply apps/console/src/features/groups/components/wizard/group-assign-users-updated.tsx]: Fixed with 921cea740ad840e2bb89d86d10c54277645d24f0

**Pull request merged**